### PR TITLE
precommit: yapf

### LINKED ISSUE #4983: use formatting Isort/YAPF/Black
## 🚀 Feature

Lets standartize the code formating:
- Isort for import cleaning
- YAPF for custom code formating
- Black is just one strong one

## Motivation

Improve the code readability

## Alternatives

could be integrated with pre-commit, see #4982

## Additional context

[Thoughts on Why Code Formatting is even more important for Open-Source](https://medium.com/geekculture/thought-on-why-code-formatting-is-important-even-more-for-open-source-476829b54eaf)

### PR DESCRIPTION
**Pull request by Borda**
use YAPF as default codebase formatting
Resolves #4983
Alternative to, closes #5490 

This PR defines YAPF as pre-commit hook and lets the bot fix it... rabbit

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Code improvements and bug fixes to enhance YOLOv5's functionalities.

### 📊 Key Changes
- 🧹 Cleaned up code by enabling yapf formatting in `.pre-commit-config.yaml`.
- ✨ Improved `detect.py` function definitions for better readability.
- 🎨 Reformatted `export.py` for readability and standardized export formats.
- 💅 Applied consistent style improvements to the codebase, such as line breaks and indentation.
- 🛠 Fixed minor issues within the code files like `hubconf.py`, `models/common.py`, and `models/experimental.py`.

### 🎯 Purpose & Impact
- 👀 These changes make the code more readable and maintainable.
- 🐛 Bug fixes prevent unexpected behavior and facilitate seamless development and deployment.
- 🔨 Enabling yapf formatting in the pre-commit workflow ensures consistent code style automatically, preventing style-related issues in future contributions.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -43,3 +43,14 @@ ignore =
     F403
     E302
     F541
+
+[yapf]
```
Comment (Borda) [inline setup.cfg:50]: @glenn-jocher here is the configuration
feel free to edit according to your preference and let the bot apply changes :rabbit:

**Comment by Borda** [comment]
well seems that the ISORT and YAPF configuration yield in inconsistent imports formatting :confused:

**Comment by glenn-jocher** [comment]
@Borda yes I see that also :(

**Comment by Borda** [comment]
> yes I see that also :(

ok, found and motion which works for both - 5 - Hanging Grid Grouped
```py
from third_party import (
    lib1, lib2, lib3, lib4,
    lib5, ...
)
```
@glenn-jocher mind checking if this is fine with you? :]

**Comment by Borda** [comment]
@glenn-jocher @kalenmike can we merge this one? :]

**Comment by Borda** [comment]
@glenn-jocher I would rather go with `multi_line_output=5` as it is more readable, the 0 is so strange...

**Comment by Borda** [comment]
I would add [Thoughts on Why Code Formatting is even more important for Open-Source](https://medium.com/geekculture/thought-on-why-code-formatting-is-important-even-more-for-open-source-476829b54eaf)

**Comment by Borda** [comment]
@glenn-jocher resolved conflicts, so shall be ready to go... :rabbit:

File: export.py [hunk 2]
```diff
@@ -104,14 +117,13 @@ def export_onnx(model, im, file, opset, train, dynamic, simplify, prefix=colorst
...
                 import onnxsim
 
                 LOGGER.info(f'{prefix} simplifying with onnx-simplifier {onnxsim.__version__}...')
-                model_onnx, check = onnxsim.simplify(
-                    model_onnx,
-                    dynamic_input_shape=dynamic,
-                    input_shapes={'images': list(im.shape)} if dynamic else None)
+                model_onnx, check = onnxsim.simplify(model_onnx,
+                                                     dynamic_input_shape=dynamic,
+                                                     input_shapes={'images': list(im.shape)} if dynamic else None)
```
Comment (Borda) [inline export.py:126]:
```suggestion
                                                     input_shapes={'images': list(im.shape)} if dynamic else None),
```

File: export.py [hunk 3]
```diff
@@ -144,9 +156,17 @@ def export_coreml(model, im, file, prefix=colorstr('CoreML:')):
...
+                       im,
+                       file,
+                       dynamic,
+                       tf_nms=False,
+                       agnostic_nms=False,
+                       topk_per_class=100,
+                       topk_all=100,
+                       iou_thres=0.45,
+                       conf_thres=0.25,
+                       prefix=colorstr('TensorFlow saved_model:')):
```
Comment (Borda) [inline export.py:169]:
```suggestion
                       prefix=colorstr('TensorFlow saved_model:'),):
```

File: export.py [hunk 4]
```diff
@@ -383,9 +402,16 @@ def run(data=ROOT / 'data/coco128.yaml',  # 'dataset.yaml path'
...
+        model = export_saved_model(model,
+                                   im,
+                                   file,
+                                   dynamic,
+                                   tf_nms=nms or agnostic_nms or tfjs,
+                                   agnostic_nms=agnostic_nms or tfjs,
+                                   topk_per_class=topk_per_class,
+                                   topk_all=topk_all,
+                                   conf_thres=conf_thres,
+                                   iou_thres=iou_thres)  # keras model
```
Comment (Borda) [inline export.py:414]:
```suggestion
                                   iou_thres=iou_thres,)  # keras model
```

File: models/common.py [hunk 5]
```diff
@@ -515,8 +520,13 @@ def forward(self, imgs, size=640, augment=False, profile=False):
...
             # Post-process
-            y = non_max_suppression(y if self.dmb else y[0], self.conf, iou_thres=self.iou, classes=self.classes,
-                                    agnostic=self.agnostic, multi_label=self.multi_label, max_det=self.max_det)  # NMS
+            y = non_max_suppression(y if self.dmb else y[0],
+                                    self.conf,
+                                    iou_thres=self.iou,
+                                    classes=self.classes,
+                                    agnostic=self.agnostic,
+                                    multi_label=self.multi_label,
+                                    max_det=self.max_det)  # NMS
```
Comment (Borda) [inline models/common.py:529]:
```suggestion
                                    max_det=self.max_det,)  # NMS
```

File: models/tf.py [hunk 6]
```diff
@@ -359,8 +373,13 @@ def predict(self, inputs, tf_nms=False, agnostic_nms=False, topk_per_class=100,
...
                 boxes = tf.expand_dims(boxes, 2)
-                nms = tf.image.combined_non_max_suppression(
-                    boxes, scores, topk_per_class, topk_all, iou_thres, conf_thres, clip_boxes=False)
+                nms = tf.image.combined_non_max_suppression(boxes,
+                                                            scores,
+                                                            topk_per_class,
+                                                            topk_all,
+                                                            iou_thres,
+                                                            conf_thres,
+                                                            clip_boxes=False)
```
Comment (Borda) [inline models/tf.py:382]:
```suggestion
                                                            clip_boxes=False,)
```

File: models/tf.py [hunk 7]
```diff
@@ -390,20 +410,26 @@ def _nms(x, topk_all=100, iou_thres=0.45, conf_thres=0.25):  # agnostic NMS
         boxes, classes, scores = x
         class_inds = tf.cast(tf.argmax(classes, axis=-1), tf.float32)
         scores_inp = tf.reduce_max(scores, -1)
-        selected_inds = tf.image.non_max_suppression(
-            boxes, scores_inp, max_output_size=topk_all, iou_threshold=iou_thres, score_threshold=conf_thres)
+        selected_inds = tf.image.non_max_suppression(boxes,
+                                                     scores_inp,
+                                                     max_output_size=topk_all,
+                                                     iou_threshold=iou_thres,
+                                                     score_threshold=conf_thres)
```
Comment (Borda) [inline models/tf.py:417]:
```suggestion
                                                     score_threshold=conf_thres,)
```

File: setup.cfg [hunk 8]
```diff
@@ -48,4 +48,17 @@ ignore =
...
+# see: https://pycqa.github.io/isort/docs/configuration/multi_line_output_modes.html
 multi_line_output = 0
+
+
+[yapf]
+based_on_style = pep8
+spaces_before_comment = 2
+# split_before_logical_operator = true
+COLUMN_LIMIT = 120
+# COALESCE_BRACKETS = true
```
Comment (Borda) [inline setup.cfg:60]:
```suggestion
COALESCE_BRACKETS = true
```

File: setup.cfg [hunk 9]
```diff
@@ -48,4 +48,17 @@ ignore =
...
+# ALLOW_SPLIT_BEFORE_DICT_VALUE = false
+# DEDENT_CLOSING_BRACKETS = true
```
Comment (Borda) [inline setup.cfg:62]:
```suggestion
DEDENT_CLOSING_BRACKETS = true
```

File: train.py [hunk 10]
```diff
@@ -210,19 +210,38 @@ def train(hyp,  # path/to/hyp.yaml or hyp dictionary
...
+                                              hyp=hyp,
+                                              augment=True,
+                                              cache=opt.cache,
+                                              rect=opt.rect,
+                                              rank=LOCAL_RANK,
+                                              workers=workers,
+                                              image_weights=opt.image_weights,
+                                              quad=opt.quad,
+                                              prefix=colorstr('train: '),
+                                              shuffle=True)
```
Comment (Borda) [inline train.py:227]:
```suggestion
                                              shuffle=True,)
```

File: train.py [hunk 11]
```diff
@@ -210,19 +210,38 @@ def train(hyp,  # path/to/hyp.yaml or hyp dictionary
...
+                                       batch_size // WORLD_SIZE * 2,
+                                       gs,
+                                       single_cls,
+                                       hyp=hyp,
+                                       cache=None if noval else opt.cache,
+                                       rect=True,
+                                       rank=-1,
+                                       workers=workers,
+                                       pad=0.5,
                                        prefix=colorstr('val: '))[0]
```
Comment (Borda) [inline train.py:245]:
```suggestion
                                       prefix=colorstr('val: '),)[0]
```

File: utils/augmentations.py [hunk 12]
```diff
@@ -121,7 +124,14 @@ def letterbox(im, new_shape=(640, 640), color=(114, 114, 114), auto=True, scaleF
...
-def random_perspective(im, targets=(), segments=(), degrees=10, translate=.1, scale=.1, shear=10, perspective=0.0,
+def random_perspective(im,
+                       targets=(),
+                       segments=(),
+                       degrees=10,
+                       translate=.1,
+                       scale=.1,
+                       shear=10,
+                       perspective=0.0,
                        border=(0, 0)):
```
Comment (Borda) [inline utils/augmentations.py:135]:
```suggestion
                       border=(0, 0),):
```

File: utils/datasets.py [hunk 13]
```diff
@@ -76,37 +76,55 @@ def exif_transpose(image):
...
+                      augment=False,
+                      cache=False,
+                      pad=0.0,
+                      rect=False,
+                      rank=-1,
+                      workers=8,
+                      image_weights=False,
+                      quad=False,
+                      prefix='',
+                      shuffle=False):
```
Comment (Borda) [inline utils/datasets.py:110]:
```suggestion
                      shuffle=False,):
```

File: utils/datasets.py [hunk 14]
```diff
@@ -377,8 +393,19 @@ class LoadImagesAndLabels(Dataset):
...
+                 batch_size=16,
+                 augment=False,
+                 hyp=None,
+                 rect=False,
+                 image_weights=False,
+                 cache_images=False,
+                 single_cls=False,
+                 stride=32,
+                 pad=0.0,
+                 prefix=''):
```
Comment (Borda) [inline utils/datasets.py:408]:
```suggestion
                 prefix='',):
```

**Comment by Borda** [comment]
@glenn-jocher any update on this format cleaning PR? :confused:

**Comment by Borda** [comment]
@glenn-jocher pls ping :rabbit: ^^

**Comment by Borda** [comment]
@glenn-jocher Happy New Year :tada:

**Comment by Borda** [comment]
@glenn-jocher all resolved, I may suggest getting it to merge and you can experiment with other formations in a follow-up PR? :)

**Comment by Borda** [comment]
@glenn-jocher all shall be fixed now :rabbit:

**Comment by Borda** [comment]
@glenn-jocher still an ongoing discussion? :rabbit:

**Comment by Borda** [comment]
@glenn-jocher never-ending resolving conflicts become quite annoying... :confused:

**Comment by Borda** [comment]
@glenn-jocher ^^

**Comment by glenn-jocher** [comment]
@Borda ok I think I have a good set of yapf arguments now. PyCharm reformat accepts the yapf reformat without conflicts.

**Comment by Borda** [comment]
> ok I think I have a good set of yapf arguments now. PyCharm reformat accepts the yapf reformat without conflicts.

perfect!

**Comment by glenn-jocher** [comment]
@Borda PR is merged. Thank you for your contributions to YOLOv5 🚀 and Vision AI ⭐

**Pull request merged**